go/vt/mysqlctl/mysqlshellbackupengine.go

```diff
@@ -0,0 +1,340 @@
+		BackupLocation: location,
+		Params:         mysqlShellLoadFlags,
+	}
+
+	data, err := json.MarshalIndent(bm, "", "  ")
+	if err != nil {
+		return BackupUnusable, vterrors.Wrapf(err, "cannot JSON encode %v", backupManifestFileName)
+	}
+	if _, err := mwc.Write([]byte(data)); err != nil {
+		return BackupUnusable, vterrors.Wrapf(err, "cannot write %v", backupManifestFileName)
+	}
+
+	params.Logger.Infof("Backup completed")
+	return BackupUsable, nil
+}
+
+func (be *MySQLShellBackupEngine) ExecuteRestore(ctx context.Context, params RestoreParams, bh backupstorage.BackupHandle) (*BackupManifest, error) {
+	params.Logger.Infof("Calling ExecuteRestore for %s (DeleteBeforeRestore: %v)", params.DbName, params.DeleteBeforeRestore)
+
+	err := be.restorePreCheck()
+	if err != nil {
+		return nil, vterrors.Wrap(err, "failed restore precheck")
+	}
+
+	var bm MySQLShellBackupManifest
+	if err := getBackupManifestInto(ctx, bh, &bm); err != nil {
+		return nil, err
+	}
+
+	// mark restore as in progress
+	if err := createStateFile(params.Cnf); err != nil {
+		return nil, err
+	}
+
+	// make sure semi-sync is disabled, otherwise we will wait forever for acknowledgements
+	err = params.Mysqld.SetSemiSyncEnabled(ctx, false, false)
+	if err != nil {
+		return nil, vterrors.Wrap(err, "disable semi-sync failed")
+	}
+
+	// if we received a RestoreFromBackup API call instead of it being a command line argument,
+	// we need to first clean the host before we start the restore.
+	if params.DeleteBeforeRestore {
+		params.Logger.Infof("restoring on an existing tablet, so dropping database %q", params.DbName)
+
+		err = params.Mysqld.ExecuteSuperQueryList(ctx,
+			[]string{fmt.Sprintf("DROP DATABASE IF EXISTS `%s`", params.DbName)},
+		)
+		if err != nil {
+			return nil, vterrors.Wrap(err, fmt.Sprintf("dropping database %q failed", params.DbName))
+		}
+
+	}
+
+	// we need to get rid of all the current replication information on the host.
+	err = params.Mysqld.ResetReplication(ctx)
+	if err != nil {
+		return nil, vterrors.Wrap(err, "unable to reset replication")
+	}
+
+	// this is required so we can load the backup generated by MySQL Shell. we will disable it afterwards.
+	err = params.Mysqld.ExecuteSuperQueryList(ctx, []string{"SET GLOBAL LOCAL_INFILE=1"})
+	if err != nil {
+		return nil, vterrors.Wrap(err, "unable to set local_infile=1")
+	}
+
+	if mysqlShellSpeedUpRestore {
+		// disable redo logging and double write buffer if we are configured to do so.
+		err = params.Mysqld.ExecuteSuperQueryList(ctx, []string{"ALTER INSTANCE DISABLE INNODB REDO_LOG"})
+		if err != nil {
+			return nil, vterrors.Wrap(err, "unable to disable REDO_LOG")
+		}
+		params.Logger.Infof("Disabled REDO_LOG")
+
+		defer func() { // re-enable once we are done with the restore.
+			err := params.Mysqld.ExecuteSuperQueryList(ctx, []string{"ALTER INSTANCE ENABLE INNODB REDO_LOG"})
+			if err != nil {
+				params.Logger.Errorf("unable to re-enable REDO_LOG: %v", err)
```

Comment:
Should we fail the restore process? I'm not sure if I have a good answer here.

Reply:
that's a good question. the original intention here is, since we were able to successfully disable Redo logs/double write buffer, it would be better to fail the restore than potentially put it in service with a potentially dangerous configuration without the user realising it. 

if the user wishes to run with redo log/double write disabled buffer they can avoid setting `--mysql_shell_speedup_restore` and handle it outside of vitess.

Reply:
Agreed, we should fail the restore when these flags are provided and if the above errors.

Reply:
Just verifying that this is still to be addressed.

Reply:
is there something you still wanted me to address? currently we will only attempt to disable/enable if the flags are passed, otherwise there is no change in behaviour and if we can't run the `ALTER INSTANCE` we will fail the backup

Reply:
> currently we will only attempt to disable/enable if the flags are passed, otherwise there is no change in behaviour and if we can't run the ALTER INSTANCE we will fail the backup

All right! Then there is nothing else to be addressed.